util/triehash-ethereum/src/lib.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Generates Keccak-flavoured trie roots.
+
+extern crate ethereum_types;
+extern crate keccak_hasher;
+extern crate triehash;
+
+use ethereum_types::H256;
+use keccak_hasher::KeccakHasher;
+
+/// Generates a trie root hash for a vector of key-value tuples
+pub fn trie_root<I, A, B>(input: I) -> H256
+where
+    I: IntoIterator<Item = (A, B)>,
+    A: AsRef<[u8]> + Ord,
+    B: AsRef<[u8]>,
+{
+    triehash::trie_root::<KeccakHasher, _, _, _>(input)
+}
+
+/// Generates a key-hashed (secure) trie root hash for a vector of key-value tuples.
+pub fn sec_trie_root<I, A, B>(input: I) -> H256
+where
+    I: IntoIterator<Item = (A, B)>,
+    A: AsRef<[u8]>,
+    B: AsRef<[u8]>,
```

Comment:
`K`/`V` makes sense, changing that. Not sure how to remove the bound though, can you show me how? Without the bound the call to `triehash::sec_trie_root()` does not work. Do you know a better way?

Reply:
well my first comment is that you could do:

```rust
pub fn sec_trie_root<I, A>(input: I) -> H256
where
    I: IntoIterator<Item = (A, A)>,
    A: AsRef<[u8]>,
{}
```
Otherwise this is also ok IMO:
```rust
pub fn sec_trie_root<I, K, V>(input: I) -> H256
where
    I: IntoIterator<Item = (K, V)>,
    K: AsRef<[u8]>,
    V: AsRef<[u8]>,
{}
```

Reply:
@niklasad1 that's not true. Right now, `A` and `B` can be different types (and they, very often are), whereas in your snippet they need to be exactly the same type

Reply:
@niklasad1 it's probably easier to see if you look at the **actual** `sec_trie_root()` method [here](https://github.com/paritytech/parity-common/blob/master/triehash/src/lib.rs#L52) – this little crate here is just a "proxy" to the real deal.

Reply:
@debris @dvdplm Marek is right even if the types are [u8; 4] and [u8; 16] it wouldn't compile because they are different types so ignore me and continue :)